### PR TITLE
FE310-based boards: transition from pinmux to pinctrl 

### PR DESCRIPTION
**Pull request by fkokosinski**
This pull request contains the following changes as part of the effort in #39740:
* introduction of the new pinctrl driver for FE310-based boards
* modifications of the existing drivers used in FE310-based boards so that they use the pinctrl API
* binding changes
* removal of the obsolete pinmux counterpart

### CODE REVIEW

File: drivers/pinctrl/Kconfig.sifive [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2022 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config PINCTRL_SIFIVE
+	bool "SiFive Freedom SoC pinmux driver"
+	depends on SOC_SERIES_RISCV_SIFIVE_FREEDOM
```
Comment (gmarull) [inline drivers/pinctrl/Kconfig.sifive:6]: please default based on `dt_compat_enabled` (see others)

File: boards/riscv/hifive1/hifive1-pinctrl.dtsi [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#include <dt-bindings/pinctrl/sifive-pinctrl.h>
+
+&pinctrl {
+	/* UART0 */
+	uart0_rx_default: uart0_rx_default {
+		pinmux = <16 SIFIVE_PINMUX_IOF0>;
+	};
+	uart0_tx_default: uart0_tx_default {
+		pinmux = <17 SIFIVE_PINMUX_IOF0>;
+	};
```
Comment (gmarull) [inline boards/riscv/hifive1/hifive1-pinctrl.dtsi:16]: suggestion: if not autogenerating nodes, grouping representation leads to a less verbose board pinctrl file.

File: boards/riscv/hifive1/hifive1_defconfig [hunk 3]
```diff
@@ -10,7 +10,8 @@ CONFIG_UART_SIFIVE_PORT_0=y
 CONFIG_UART_CONSOLE=y
 CONFIG_PLIC=y
 CONFIG_PINMUX=y
-CONFIG_PINMUX_SIFIVE=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_SIFIVE=y
```
Comment (gmarull) [inline boards/riscv/hifive1/hifive1_defconfig:14]: only `CONFIG_PINMUX=y` should be required, see my previous comment about defaulting based on dt_compat_status. If all boards need this flag, consider enabling this at SoC level.

File: dts/bindings/spi/sifive,spi0.yaml [hunk 4]
```diff
@@ -5,8 +5,14 @@ description: Sifive SPI controller
...
 
 properties:
     reg:
       required: true
+
+    pinctrl-0:
+      required: false
+
+    pinctrl-names:
+      required: false
```
Comment (gmarull) [inline dts/bindings/spi/sifive,spi0.yaml:18]: this is redundant, none are required by default. applies to others as well.

File: boards/riscv/hifive1/hifive1_defconfig [hunk 5]
```diff
@@ -10,7 +10,8 @@ CONFIG_UART_SIFIVE_PORT_0=y
 CONFIG_UART_CONSOLE=y
 CONFIG_PLIC=y
 CONFIG_PINMUX=y
```
Comment (gmarull) [inline boards/riscv/hifive1/hifive1_defconfig:12]: pinmux driver seems to be removed, so it should be this option

**Review by gmarull — CHANGES_REQUESTED** [review]
Thanks for the contribution, I have a few comments (some non-blocking)

File: dts/bindings/spi/sifive,spi0.yaml (hunk 4 above)
Comment (fkokosinski) [reply dts/bindings/spi/sifive,spi0.yaml]: Removed redundant fields from binding yamls

File: drivers/pinctrl/Kconfig.sifive (hunk 1 above)
Comment (fkokosinski) [reply drivers/pinctrl/Kconfig.sifive:8]: Done

File: boards/riscv/hifive1/hifive1_defconfig (hunk 3 above)
Comment (fkokosinski) [reply boards/riscv/hifive1/hifive1_defconfig]: Removed `CONFIG_PINCTRL_SIFIVE=y`

File: dts/bindings/pinctrl/sifive,pinctrl.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2022 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  SiFive FE310 IO Function (iof)
+  Binding covers the IOF_EN/IOF_SEL registers that are a subset of the GPIO
+  controller.
```
Comment (gmarull) [inline dts/bindings/pinctrl/sifive,pinctrl.yaml:7]: nitpick: please extend this documentation on how pinctrl is expected to be used (see other vendors as a reference)

File: boards/riscv/hifive1/hifive1_defconfig (hunk 5 above)
Comment (fkokosinski) [reply boards/riscv/hifive1/hifive1_defconfig]: Done.

File: dts/bindings/pinctrl/sifive,pinctrl.yaml (hunk 6 above)
Comment (fkokosinski) [reply dts/bindings/pinctrl/sifive,pinctrl.yaml]: Done.

File: drivers/pinctrl/pinctrl_sifive.c [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	if (func == SIFIVE_PINMUX_IOF1) {
+		val |= (SIFIVE_PINMUX_IOF1 << pin);
+	} else {
+		val &= ~(SIFIVE_PINMUX_IOF1 << pin);
+	}
+	sys_write32(val, PINCTRL_IOF_SEL);
+
+	/* Enable IO function for this pin */
+	val = sys_read32(PINCTRL_IOF_EN);
+	val |= (1 << pin);
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_sifive.c:35]:
```suggestion
	val |= BIT(pin);
```

File: drivers/serial/uart_sifive.c [hunk 8]
```diff
@@ -334,6 +339,13 @@ static int uart_sifive_init(const struct device *dev)
 	cfg->cfg_func();
 #endif
 
+#ifdef CONFIG_PINCTRL
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+#endif
```
Comment (gmarull) [inline drivers/serial/uart_sifive.c:347]: does driver work at all without pinctrl? why is this made optional (same applies to others)?

Comment (fkokosinski) [reply drivers/serial/uart_sifive.c:347]: This change was necessary for the `hifive_unleashed` and `hifive_unmatched` boards, as they weren't using the old pinmux driver before.

File: drivers/pinctrl/pinctrl_sifive.c (hunk 7 above)
Comment (fkokosinski) [reply drivers/pinctrl/pinctrl_sifive.c]: Done.

File: drivers/serial/uart_sifive.c (hunk 8 above)
Comment (fkokosinski) [reply drivers/serial/uart_sifive.c:347]: They don't do it at all, if I'm correct. I've re-checked this with their SoCs manuals (FU540 and FU740 respectively), and the section that would describe IOFs is empty.

**Review by gmarull — APPROVED** [review]

**Review by tgorochowik — APPROVED** [review]

**Pull request merged**